core/src/main/java/org/elasticsearch/search/aggregations/support/InternalOrder.java

```diff
@@ -0,0 +1,603 @@
+            CompoundOrder other = (CompoundOrder) obj;
+            return Objects.equals(orderElements, other.orderElements);
+        }
+
+        /**
+         * {@code Comparator} for sorting buckets by multiple criteria.
+         */
+        static class CompoundOrderComparator implements Comparator<Bucket> {
+
+            private List<BucketOrder> compoundOrder;
+            private Aggregator aggregator;
+
+            /**
+             * Create a new {@code Comparator} for sorting buckets by multiple criteria.
+             *
+             * @param compoundOrder a list of {@link BucketOrder}s to sort on, in order of priority.
+             * @param aggregator    {@link BucketOrder#comparator(Aggregator)}
+             */
+            CompoundOrderComparator(List<BucketOrder> compoundOrder, Aggregator aggregator) {
+                this.compoundOrder = compoundOrder;
+                this.aggregator = aggregator;
+            }
+
+            @Override
+            public int compare(Bucket b1, Bucket b2) {
+                int result = 0;
+                for (Iterator<BucketOrder> itr = compoundOrder.iterator(); itr.hasNext() && result == 0; ) {
+                    result = itr.next().comparator(aggregator).compare(b1, b2);
+                }
+                return result;
+            }
+        }
+    }
+
+    private static final byte COUNT_DESC_ID = 1;
+    private static final byte COUNT_ASC_ID = 2;
+    private static final byte KEY_DESC_ID = 3;
+    private static final byte KEY_ASC_ID = 4;
+
+    /**
+     * Order by the (higher) count of each bucket.
+     */
+    static final InternalOrder COUNT_DESC = new InternalOrder(COUNT_DESC_ID, "_count", false, comparingCounts().reversed());
+
+    /**
+     * Order by the (lower) count of each bucket.
+     */
+    static final InternalOrder COUNT_ASC = new InternalOrder(COUNT_ASC_ID, "_count", true, comparingCounts());
+
+    /**
+     * Order by the key of each bucket descending.
+     */
+    static final InternalOrder KEY_DESC = new InternalOrder(KEY_DESC_ID, "_key", false, comparingKeys().reversed());
+
+    /**
+     * Order by the key of each bucket ascending.
+     */
+    static final InternalOrder KEY_ASC = new InternalOrder(KEY_ASC_ID, "_key", true, comparingKeys());
+
+    /**
+     * @return compare by {@link Bucket#getDocCount()}.
+     */
+    private static Comparator<Bucket> comparingCounts() {
+        return Comparator.comparingLong(Bucket::getDocCount);
+    }
+
+    /**
+     * @return compare by {@link Bucket#getKey()} from the appropriate implementation.
+     */
+    private static Comparator<Bucket> comparingKeys() {
+        return (b1, b2) -> {
+            if (b1 instanceof Terms.Bucket) {
+                return ((Terms.Bucket) b1).compareTerm((Terms.Bucket) b2);
+            } else if (b1 instanceof InternalHistogram.Bucket) {
+                return Double.compare(((InternalHistogram.Bucket) b1).key, ((InternalHistogram.Bucket) b2).key);
```

Comment:
Yup that would work, however this will make this code pretty ugly with even more casts. I'm starting to really like your original suggestion of making the buckets `implements Comparable`.